### PR TITLE
Add test coverage to library and s3 stubbing

### PR DESCRIPTION
**Pull request by rymarczy**
Almost 100% coverage for `py_gtfs_rt_ingestion` library files as reported by pytest-cov.
- add `pytest-cov` as poetry dev dependency
- update `s3_utils.py` to use boto3 client instead of resource
- initial attempt at S3 stubbing in `test_s3_utils.py`
- various additions to library test files for near 100% coverage as reported by pytset-cov

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,4 +12,6 @@ erl_crash.dump
 
 __pycache__
 venv
-*.ipynb
\ No newline at end of file
+*.ipynb
+.coverage
+htmlcov
```
Comment (rymarczy) [inline .gitignore:17]: additional ignores from pytest-cov

File: py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/convert.py [hunk 2]
```diff
@@ -15,7 +15,7 @@ def gz_to_pyarrow(filename: str, config: Configuration):
     """
     logging.info("Converting %s to Parquet Table" % filename)
     try:
-        if str(filename).startswith('s3'):
+        if str(filename).startswith('s3://'):
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/convert.py:18]: might as well check for for s3 starting path.

File: py_gtfs_rt_ingestion/tests/test_batcher.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    files = [
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', 100_000),
+    ]
+    batches = [b for b in batch_files(files=files, threshold=1)]
+    assert  len(batches) == 1
+
+    files = [
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', 1_000),
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', 1_000),
+    ]
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:55]: I am open to comments on any of this, still figuring out all this testing stuff.

File: py_gtfs_rt_ingestion/tests/test_configuration.py [hunk 4]
```diff
@@ -31,15 +32,13 @@ def test_filname_parsing():
     with pytest.raises(ConfigTypeFromFilenameException):
         ConfigType.from_filename('this.is.a.bad.filename.json.gz')
 
+    with pytest.raises(NoImplException):
+        Configuration('this.is.a.bad.filename.json.gz')
+
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_configuration.py:37]: Check Exception for initializing `Configuration` object with bad file name.

File: py_gtfs_rt_ingestion/tests/test_configuration.py [hunk 5]
```diff
@@ -31,15 +32,13 @@ def test_filname_parsing():
...
 def test_get_schema():
     config = Configuration(filename=vehicle_positions_filename)
     schema = config.export_schema
 
-    # Requires full implemenation of ConfigDetail for ConfigType
-    # with pytest.raises(Exception):
-    #     unimpled_config = Configuration.RT_ALERTS
-    #     unimpled_schema = unimpled_config.get_schema()
-
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_configuration.py:42]: Don't think this is needed anymore.

File: py_gtfs_rt_ingestion/tests/test_s3_utils.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        "Prefix": ANY,
+    }
+    page_obj_response = {
+        'KeyCount': 0,
+        'Contents': [],
+    }
+    s3_stub.add_response("list_objects_v2", page_obj_response, page_obj_params  )
+    with s3_stub:
+        files = [file for file in file_list_from_s3('mbta-gtfs-s3','')]
+        assert files == []
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_s3_utils.py:53]: Very open to comments on this, just my initial attempt at stubbing S3 calls.

File: py_gtfs_rt_ingestion/tests/test_convert.py [hunk 7]
```diff
@@ -1,9 +1,37 @@
...
+    empty_file = os.path.join(TEST_FILE_DIR, "empty.json.gz")
+    table = gz_to_pyarrow(filename=empty_file, config=config)
+    np_df = table.to_pandas()
+    assert np_df.shape == (0,len(config.export_schema))
+
+    one_blank_file = os.path.join(TEST_FILE_DIR, "one_blank_record.json.gz")
+    table = gz_to_pyarrow(filename=one_blank_file, config=config)
+    np_df = table.to_pandas()
+    assert np_df.shape == (1,len(config.export_schema))
+
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_convert.py:34]: Add testing for dummy json.gz files with 0 records and 1 record of empty data. Hopefully covers edge cases...

File: py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/batcher.py [hunk 8]
```diff
@@ -64,7 +64,7 @@ def batch_files(files: Iterable[(str, int)], threshold: int) -> Iterable[Batch]:
 
         batch = ongoing_batches[config_type]
 
-        if batch.total_size + int(size) > threshold:
+        if batch.total_size + int(size) > threshold and len(batch.filenames) > 0:
```
[the remark this answers is not in the document]
Comment (mzappitello) [reply py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/batcher.py:67]: 👍

File: py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py [hunk 9]
```diff
@@ -15,8 +18,14 @@ def file_list_from_s3(bucket_name: str,
...
+    paginator = s3_client.get_paginator('list_objects_v2')
+    pages = paginator.paginate(
+        Bucket=bucket_name,
+        Prefix=file_prefix,
+    )
+    for page in pages:
+        if page['KeyCount'] == 0:
+            continue
+        for obj in page['Contents']:
+            yield (obj['Key'], obj['Size'])
```
[the remark this answers is not in the document]
Comment (mzappitello) [reply py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py:31]: sounds good.

File: py_gtfs_rt_ingestion/tests/test_batcher.py (hunk 3 above)
Comment (mzappitello) [reply py_gtfs_rt_ingestion/tests/test_batcher.py:55]: it feels like its tough to do elegantly.

i think i would break up this test into different concerns.
* a test to check how bad filenames are handled, including some tests where there are a mix of good and bad names.
* a test to check that small thresholds work with various file sizes.
* a test to check that we can do this on a large scale using a test file that has a bunch of files listed and that thresholds are respected

File: py_gtfs_rt_ingestion/tests/test_convert.py [hunk 10]
```diff
@@ -44,7 +81,7 @@ def test_vehicle_positions_file_conversion(tmpdir):
         'vehicle_consist':(324,'object','nan','nan'),
     }
 
-    assert np_df.shape == (426,25)
+    assert np_df.shape == (426, len(config.export_schema))
```
Comment (mzappitello) [inline py_gtfs_rt_ingestion/tests/test_convert.py:84]: probably worth a comment where the 426 comes from.

File: py_gtfs_rt_ingestion/tests/test_convert.py [hunk 11]
```diff
@@ -106,7 +146,7 @@ def test_rt_alert_file_conversion(tmpdir):
         'informed_entity':(0,'object','nan','nan'),
     }
 
-    assert np_df.shape == (144,28)
+    assert np_df.shape == (144, len(config.export_schema))
```
Comment (mzappitello) [inline py_gtfs_rt_ingestion/tests/test_convert.py:149]: also worth a comment.

**Comment by mzappitello** [comment]
a few comments here and there, but overall LGTM.

it would be nice to eventually get that coverage info posted into our actions.

it also feels like a good idea to fail the tests if we fall below a certain coverage amount. that way, we wouldn't be able to add new features without testing them.

but both of those things can go in at a future date.

File: py_gtfs_rt_ingestion/tests/test_convert.py (hunk 11 above)
Comment (rymarczy) [reply py_gtfs_rt_ingestion/tests/test_convert.py:151]: comment added

File: py_gtfs_rt_ingestion/tests/test_convert.py [hunk 12]
```diff
@@ -159,7 +202,7 @@ def test_rt_trip_file_conversion(tmpdir):
         'vehicle_label':(55,'object','nan','nan'),
     }
 
-    assert np_df.shape == (79,17)
+    assert np_df.shape == (79, len(config.export_schema))
```
[the remark this answers is not in the document]
Comment (rymarczy) [reply py_gtfs_rt_ingestion/tests/test_convert.py:208]: comment added

File: py_gtfs_rt_ingestion/tests/test_s3_utils.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    with open(large_response_file, 'r') as f:
+        page_obj_response = json.load(f)
+    s3_stub.add_response("list_objects_v2", page_obj_response, page_obj_params)
+    with s3_stub:
+        files = [file for file in file_list_from_s3('mbta-gtfs-s3','')]
+
+    assert len(files) == page_obj_response['KeyCount']
+    assert files == [(d['Key'], d['Size']) for d in page_obj_response['Contents']]
+
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_s3_utils.py:80]: Added `test_files/large_page_obj_response.json` as obfuscated s3 `list_objects` results.

File: py_gtfs_rt_ingestion/tests/test_batcher.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    s3_stub.add_response("list_objects_v2", page_obj_response, page_obj_params)
+    with s3_stub:
+        files = [file for file in file_list_from_s3('mbta-gtfs-s3','')]
+
+    batches = [b for b in batch_files(files=files, threshold=threshold)]
+    # Verify expected number of batches produced
+    assert len(batches) == 20
+    # Verify each batch total_size respects threshold
+    for b in batches:
+        assert b.total_size <= threshold
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:99]: Test `batch_files` with large json response object. Verify all batches are less than or equal to `threshold` size.

File: py_gtfs_rt_ingestion/tests/test_batcher.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    assert len(batches) == len(files)
+
+    files = [
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', threshold),
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', threshold),
+    ]
+    batches = [b for b in batch_files(files=files, threshold=threshold)]
+    assert len(batches) == len(files)
+    for b in batches:
+        assert b.total_size == threshold
```
Comment (rymarczy) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:79]: Break up testing of batches into separate functions.

File: py_gtfs_rt_ingestion/tests/test_s3_utils.py (hunk 6 above)
Comment (rymarczy) [reply py_gtfs_rt_ingestion/tests/test_s3_utils.py]: Larger `page_obj_response` saved as json test file.

File: py_gtfs_rt_ingestion/tests/test_batcher.py (hunk 3 above)
Comment (rymarczy) [reply py_gtfs_rt_ingestion/tests/test_batcher.py]: Broke up this file into a few different functions.

File: py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py [hunk 16]
```diff
@@ -15,8 +18,14 @@ def file_list_from_s3(bucket_name: str,
...
-        yield (bucket_object.key, bucket_object.size)
+    s3_client = get_s3_client()
+    paginator = s3_client.get_paginator('list_objects_v2')
+    pages = paginator.paginate(
+        Bucket=bucket_name,
+        Prefix=file_prefix,
+    )
+    for page in pages:
+        if page['KeyCount'] == 0:
+            continue
```
Comment (mzappitello) [inline py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py:29]: is this necessary? if the key count is 0, isn't `page['Contents']` just an empty list?

File: py_gtfs_rt_ingestion/tests/test_batcher.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    assert [b for b in batch_files(files=[], threshold=100)] == []
+
+def test_batch_files(s3_stub):
+    threshold = 100_000
+
+    files = [
+        ('https_cdn.mbta.com_realtime_VehiclePositions_enhanced.json.gz', 100_000),
+    ]
+    batches = [b for b in batch_files(files=files, threshold=threshold)]
+    assert len(batches) == len(files)
```
Comment (mzappitello) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:70]: this should probably be `len(batches) == 1`.

File: py_gtfs_rt_ingestion/tests/test_batcher.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    with open(large_response_file, 'r') as f:
+        page_obj_response = json.load(f)
+    s3_stub.add_response("list_objects_v2", page_obj_response, page_obj_params)
+    with s3_stub:
+        files = [file for file in file_list_from_s3('mbta-gtfs-s3','')]
+
+    batches = [b for b in batch_files(files=files, threshold=threshold)]
+    # Verify expected number of batches produced
+    assert len(batches) == 20
```
Comment (mzappitello) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:96]: i'm assuming you got this from the initial run of the batching script?

it might be a good idea to run this with a handful of threshold values and see that the number of batches goes up as the threshold goes down.

you can also test batch total size against the threshold for each value.

File: py_gtfs_rt_ingestion/tests/test_batcher.py (hunk 14 above)
Comment (mzappitello) [inline py_gtfs_rt_ingestion/tests/test_batcher.py:99]: this should probably be wrapped in an `if len(b.files) > 1` gate, since we decided to allow batches to contain a single file if they exceed threshold size.

File: py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py (hunk 16 above)
Comment (rymarczy) [reply py_gtfs_rt_ingestion/py_gtfs_rt_ingestion/s3_utils.py:29]: In my testing, `KeyCount` is always in the response object. If `KeyCount == 0` then a `Contents` object is not sent.

**Comment by mzappitello** [comment]
LGTM. Don't forget to Squash and Merge.
